daffodil-runtime1/src/main/scala/org/apache/daffodil/dpath/FNFunctions.scala

```diff
@@ -803,6 +805,85 @@ case class FNFloor(recipe: CompiledDPath, argType: NodeInfo.Kind) extends FNOneA
   }
 }
 
+/**
+ * This is "arithmetic" shift left, meaning it sign-extends a negative signed value.
+ * dfdl:leftShift for signed values does sign-extension, and if you want logical (zero-filling) left shift you must use an unsigned type
+ *
+ * @param recipes
+ * @param argType
+ */
+case class DFDLXLeftShift(recipes: List[CompiledDPath], argType: NodeInfo.Kind) extends FNTwoArgs(recipes)  {
+  override def computeValue(arg1: DataValuePrimitive, arg2: DataValuePrimitive, dstate: DState): DataValuePrimitive = {
+
+    val shiftLong = arg2.getLong
+    val shift = shiftLong.toInt
+    val width = argType.asInstanceOf[PrimNumeric].width.get
+    Assert.invariant(shift >= 0)
+    if(shift>=width)
+      dstate.SDE("Left shift not supported for shift greater or equal to "+ width)
+
+    argType match {
+      case NodeInfo.Long => arg1.getLong << shift
+
+      case NodeInfo.Int => arg1.getInt << shift
+
+      case NodeInfo.Short => (arg1.getShort << shift).toShort
+
+      case NodeInfo.Byte => (arg1.getByte << shift).toByte
+
+      case NodeInfo.UnsignedLong =>arg1.getBigInt.shiftLeft(shift)
+
+      case NodeInfo.UnsignedInt => ULong(arg1.getLong << shift).toLong
+
+      case NodeInfo.UnsignedShort => UInt(arg1.getInt << shift).toInt
+
+      case NodeInfo.UnsignedByte => UInt(arg1.getShort << shift).toShort
```

Comment:
I would prefer to see all these empty lines removed between these cases. Standard practice in scala is to not have this extra whitespace.